File: include/PolyWorksSDK/samples/plugins/common/VisualBasic.NET/VBNetCommands/COM_SDK/CCompanyNameCommands.vb
' =================================================================================================
' Plug-in's commands.
'
' Copyright © InnovMetric Logiciels Inc. 2011 All Rights Reserved
' -------------------------------------------------------------------------------------------------
Namespace VBNetCommands

    ' =============================================================================================
    ' Settings
    ' =============================================================================================
    Public Class Settings

        Public Shared g_doubleSettingValue As Double = 110.25
        Public Shared g_integerSettingValue As Integer = 237
        Public Shared g_stringSettingValue As String = "String Value"

    End Class

    Public Class CCmdCompanyNameCmdFromMacro
        Inherits CComCommand

        Private m_inspectCmdCenter As IMInspectLib.IIMCommandCenter

        ' =========================================================================================
        ' Command constructor. 
        ' This constructor is only called one time when the command is registered to the 
        ' application.
        '
        ' Parameter: info_                 : Parameters necessary to build the command.
        ' Parameter: inspectCommandCenter_ : Command center to execute the macro script.
        ' -----------------------------------------------------------------------------------------
        Public Sub New(ByVal info_ As CComCommandInfo, ByRef inspectCommandCenter_ As IMInspectLib.IIMCommandCenter)

            MyBase.New(info_)
            m_inspectCmdCenter = inspectCommandCenter_

        End Sub

        ' =========================================================================================
        ' Command action, which executes a macro script on the host application. 
        '
        ' TODO: Comment the action here
        '
        ' Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        '
        ' Return: COM error code.
        ' -----------------------------------------------------------------------------------------
        Public Overrides Function Execute2(ByVal pIMCommandArgumentEnum_ As IMPluginLib.IIMCommandArgumentEnum) As IMPluginLib.ECommandReturnCodes

            Dim returnValue As Integer
            Dim script As String

            ' Build the macro script
            ' TODO: Replace macro script here.

            script = "version ""5.0""" & vbCrLf

            ' Turn Off the echoing of the commands
            script += "MACRO COMMAND ECHO ( ""Off"" )" & vbCrLf

            ' We want to keep all measurement objects visible while probing to be able to move the probe on the plane (see below)
            script += "PROBE DISPLAY SHOW_ONLY_RELEVANT_MEASUREMENT_OBJECTS ( ""Off"" )" & vbCrLf

            ' Initialisation
            ' Create a plane, and a circle to be probed
            script += "FILE NEW_PROJECT" & vbCrLf &
                "FEATURE PRIMITIVE PLANE CREATE FROM_POINT_AND_NORMAL ( 43.882238437618, 6.96668688420377, 1.2, 0.0, 0.0, 1.0, ""Nominal"", ""plane 1"", , 100, 4 )" & vbCrLf &
                "TREEVIEW FEATURE Select ( 1, ""On"" )" & vbCrLf &
                "TREEVIEW VIEW STATIC SMOOTH" & vbCrLf &
                "FEATURE PRIMITIVE CIRCLE CREATE ( 25.3999999904043, 50.7999999963512, 0.0, 0.0, 0.0, 1.0, 12.6999997329769, ""Nominal"", ""circle 1"" )" & vbCrLf

            ' Probe circle using the Virtual Probe.
            ' The user can move the probe on the previously created plane and use the Insert key to acquire points.
            ' The execution of the script will wait until the probing Is completed before continuing.
            script += "" & vbCrLf &
                "PROBE DEVICE ( ""PolyWorks Virtual Probe"" )" & vbCrLf &
                "FEATURE PRIMITIVE CIRCLE PROBE ( ""Measured"", ""circle 1"" )" & vbCrLf

            ' Create and open a formatted report
            script += "TREEVIEW FEATURE Select ( 1, ""On"" )" & vbCrLf &
                "REPORT_ITEM TABLE FROM_SELECTED_OBJECTS ( )" & vbCrLf &
                "TREEVIEW OBJECT SELECT NONE" & vbCrLf &
                "TREEVIEW FORMATTED_REPORT Select ( 1, ""On"" )" & vbCrLf &
                "FORMATTED_REPORT OPEN" & vbCrLf

            ' Reset the option we changed previously to the default value
            script += "PROBE DISPLAY SHOW_ONLY_RELEVANT_MEASUREMENT_OBJECTS ( ""On"" )" & vbCrLf

            ' Turn back On the echoing of the commands
            script += "MACRO COMMAND ECHO ( ""On"" )" & vbCrLf

            ' Execute the macro script
            returnValue = m_inspectCmdCenter.ScriptExecuteFromBuffer(script, "")

            ' Check if the execution is a success
            If m_inspectCmdCenter.ReturnValueIsSuccess(returnValue) = 1 Then

                Return IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED

            End If

            Return IMPluginLib.ECommandReturnCodes.E_CMD_FAILED

        End Function

    End Class

    ' =============================================================================================
    ' TODO: Rename CompanyNameAction with your company name and your action
    '
    ' TODO: Add command description here.
    ' ---------------------------------------------------------------------------------------------
    Public Class CCmdCompanyNameAction
        Inherits CComCommand

        ' =========================================================================================
        ' Command constructor. 
        ' This constructor is only called one time when the command is registered to the 
        ' application.
        '
        ' Parameter: info_ : Parameters necessary to build the command.
        ' -----------------------------------------------------------------------------------------
        Public Sub New(ByVal info_ As CComCommandInfo)

            MyBase.New(info_)

        End Sub

        ' =========================================================================================
        ' Command action. 
        '
        ' TODO: Comment the action here
        '
        ' Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        '
        ' Return: COM error code.
        ' -----------------------------------------------------------------------------------------
        Public Overrides Function Execute2(ByVal pIMCommandArgumentEnum_ As IMPluginLib.IIMCommandArgumentEnum) As IMPluginLib.ECommandReturnCodes

            ' TODO: Add command execution code here.

            Return IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED

        End Function

    End Class

    ' =============================================================================================
    ' TODO: Rename CompanyNameDouble with your company name and double setting name.
    '
    ' Sets a double value setting within the plug-in
    ' ---------------------------------------------------------------------------------------------
    Public Class CCmdCompanyNameDoubleSet
        Inherits CComCommand

        ' =========================================================================================
        ' Command constructor. 
        ' This constructor is only called one time when the command is registered to the 
        ' application.
        '
        ' Parameter: info_ : Parameters necessary to build the command.
        ' -----------------------------------------------------------------------------------------
        Public Sub New(ByVal info_ As CComCommandInfo)

            MyBase.New(info_)

        End Sub

        ' =========================================================================================
        ' Sets a double value setting within the plug-in.
        '
        ' Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        '
        ' Return: COM error code.
        ' -----------------------------------------------------------------------------------------
        Public Overrides Function Execute2(ByVal pIMCommandArgumentEnum_ As IMPluginLib.IIMCommandArgumentEnum) As IMPluginLib.ECommandReturnCodes

            Dim retCode As IMPluginLib.ECommandReturnCodes = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED

            ' Getting argument value
            Dim success As Boolean = False
            Dim argValue As Double = 0.0

            ' Setting argument value
            If Not pIMCommandArgumentEnum_ Is Nothing Then

                ' Checking argument count validity
                If pIMCommandArgumentEnum_.ArgsGetNb() > 0 Then

                    ' Checking availability of first argument
                    Dim argument As IMPluginLib.IIMCommandArgument = Nothing
                    pIMCommandArgumentEnum_.ArgGet(0, argument)

                    If Not argument Is Nothing Then

                        ' Checking and retrieving argument value if of valid type
                        If argument.ArgValueGetAsDouble(1, argValue) = 1 Then

                            success = True

                        End If

                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument)

                    End If

                End If

            End If

            ' Argument value validation
            If success = True Then

                ' TODO: Replace validation code by your own depending on wanted argument value valid range
                If argValue < 0 Then

                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_ARG_OUT_OF_RANGE

                Else

                    ' TODO: Register setting value into the plug-in for future use or execute action using argument value
                    ' TODO: Replace g_doubleSettingValue use by your own code
                    Settings.g_doubleSettingValue = argValue

                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED

                End If

            End If

            System.Runtime.InteropServices.Marshal.ReleaseComObject(pIMCommandArgumentEnum_)

            Return retCode

        End Function

    End Class

    ' =============================================================================================
    ' TODO: Rename CompanyNameDouble with your company name and double setting name.
    '
    ' Gets a double value setting from the plug-in
    ' ---------------------------------------------------------------------------------------------
    Public Class CCmdCompanyNameDoubleGet
        Inherits CComCommand

        ' =========================================================================================
        ' Command constructor. 
        ' This constructor is only called one time when the command is registered to the 
        ' application.
        '
        ' Parameter: info_ : Parameters necessary to build the command.
        ' -----------------------------------------------------------------------------------------
        Public Sub New(ByVal info_ As CComCommandInfo)

            MyBase.New(info_)

        End Sub

        ' =========================================================================================
        ' Gets a double value setting from the plug-in
        '
        ' Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        '
        ' Return: COM error code.
        ' -----------------------------------------------------------------------------------------
        Public Overrides Function Execute2(ByVal pIMCommandArgumentEnum_ As IMPluginLib.IIMCommandArgumentEnum) As IMPluginLib.ECommandReturnCodes

            Dim retCode As IMPluginLib.ECommandReturnCodes = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED

            ' Getting argument value
            If Not pIMCommandArgumentEnum_ Is Nothing Then

                ' Checking argument count validity
                If pIMCommandArgumentEnum_.ArgsGetNb() > 0 Then

                    ' Checking availability of first argument
                    Dim argument As IMPluginLib.IIMCommandArgument = Nothing
                    pIMCommandArgumentEnum_.ArgGet(0, argument)

                    If Not argument Is Nothing Then

                        ' TODO: Get setting value from the plug-in
                        ' TODO: Replace g_doubleSettingValue use by your own code
                        If argument.ArgValueSetFromDouble(1, Settings.g_doubleSettingValue) = 1 Then

                            retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED

                        End If

                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument)

                    End If

                End If

            End If

            System.Runtime.InteropServices.Marshal.ReleaseComObject(pIMCommandArgumentEnum_)

            Return retCode

        End Function

    End Class

    ' =============================================================================================
    ' TODO: Rename CompanyNameInteger with your company name and integer setting name.
    '
    ' Sets a integer value setting within the plug-in
    ' ---------------------------------------------------------------------------------------------
    Public Class CCmdCompanyNameIntegerSet
        Inherits CComCommand

        ' =========================================================================================
        ' Command constructor. 
        ' This constructor is only called one time when the command is registered to the 
        ' application.
        '
        ' Parameter: info_ : Parameters necessary to build the command.
        ' -----------------------------------------------------------------------------------------
        Public Sub New(ByVal info_ As CComCommandInfo)

            MyBase.New(info_)

        End Sub

        ' =========================================================================================
        ' Sets a integer value setting within the plug-in.
        '
        ' Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        '
        ' Return: COM error code.
        ' -----------------------------------------------------------------------------------------
        Public Overrides Function Execute2(ByVal pIMCommandArgumentEnum_ As IMPluginLib.IIMCommandArgumentEnum) As IMPluginLib.ECommandReturnCodes

            Dim retCode As IMPluginLib.ECommandReturnCodes = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED

            ' Getting argument value
            Dim success As Boolean = False
            Dim argValue As Integer = 0

            ' Setting argument value
            If Not pIMCommandArgumentEnum_ Is Nothing Then

                ' Checking argument count validity
                If pIMCommandArgumentEnum_.ArgsGetNb() > 0 Then

                    ' Checking availability of first argument
                    Dim argument As IMPluginLib.IIMCommandArgument = Nothing
                    pIMCommandArgumentEnum_.ArgGet(0, argument)

                    If Not argument Is Nothing Then

                        ' Checking and retrieving argument value if of valid type
                        If argument.ArgValueGetAsInt(1, argValue) = 1 Then

                            success = True

                        End If

                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument)

                    End If

                End If

            End If

            ' Argument value validation
            If success = True Then

                ' TODO: Replace validation code by your own depending on wanted argument value valid range
                If argValue < 0 Then

                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_ARG_OUT_OF_RANGE

                Else

                    ' TODO: Register setting value into the plug-in for future use or execute action using argument value
                    ' TODO: Replace g_integerSettingValue use by your own code
                    Settings.g_integerSettingValue = argValue

                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED

                End If

            End If

            System.Runtime.InteropServices.Marshal.ReleaseComObject(pIMCommandArgumentEnum_)

            Return retCode

        End Function

    End Class

    ' =============================================================================================
    ' TODO: Rename CompanyNameInteger with your company name and integer setting name.
    '
    ' Gets a integer value setting from the plug-in
    ' ---------------------------------------------------------------------------------------------
    Public Class CCmdCompanyNameIntegerGet
        Inherits CComCommand

        ' =========================================================================================
        ' Command constructor. 
        ' This constructor is only called one time when the command is registered to the 
        ' application.
        '
        ' Parameter: info_ : Parameters necessary to build the command.
        ' -----------------------------------------------------------------------------------------
        Public Sub New(ByVal info_ As CComCommandInfo)

            MyBase.New(info_)

        End Sub

        ' =========================================================================================
        ' Gets a integer value setting from the plug-in
        '
        ' Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        '
        ' Return: COM error code.
        ' -----------------------------------------------------------------------------------------
        Public Overrides Function Execute2(ByVal pIMCommandArgumentEnum_ As IMPluginLib.IIMCommandArgumentEnum) As IMPluginLib.ECommandReturnCodes

            Dim retCode As IMPluginLib.ECommandReturnCodes = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED

            ' Getting argument value
            If Not pIMCommandArgumentEnum_ Is Nothing Then

                ' Checking argument count validity
                If pIMCommandArgumentEnum_.ArgsGetNb() > 0 Then

                    ' Checking availability of first argument
                    Dim argument As IMPluginLib.IIMCommandArgument = Nothing
                    pIMCommandArgumentEnum_.ArgGet(0, argument)

                    If Not argument Is Nothing Then

                        ' TODO: Get setting value from the plug-in
                        ' TODO: Replace g_integerSettingValue use by your own code
                        If argument.ArgValueSetFromInt(1, Settings.g_integerSettingValue) = 1 Then

                            retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED

                        End If

                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument)

                    End If

                End If

            End If

            System.Runtime.InteropServices.Marshal.ReleaseComObject(pIMCommandArgumentEnum_)

            Return retCode

        End Function

    End Class

    ' =============================================================================================
    ' TODO: Rename CompanyNameString with your company name and string setting name.
    '
    ' Sets a string value setting within the plug-in
    ' ---------------------------------------------------------------------------------------------
    Public Class CCmdCompanyNameStringSet
        Inherits CComCommand

        ' =========================================================================================
        ' Command constructor. 
        ' This constructor is only called one time when the command is registered to the 
        ' application.
        '
        ' Parameter: info_ : Parameters necessary to build the command.
        ' -----------------------------------------------------------------------------------------
        Public Sub New(ByVal info_ As CComCommandInfo)

            MyBase.New(info_)

        End Sub

        ' =========================================================================================
        ' Sets a string value setting within the plug-in.
        '
        ' Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        '
        ' Return: COM error code.
        ' -----------------------------------------------------------------------------------------
        Public Overrides Function Execute2(ByVal pIMCommandArgumentEnum_ As IMPluginLib.IIMCommandArgumentEnum) As IMPluginLib.ECommandReturnCodes

            Dim retCode As IMPluginLib.ECommandReturnCodes = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED

            ' Getting argument value
            Dim success As Boolean = False
            Dim argValue As String = String.Empty

            ' Setting argument value
            If Not pIMCommandArgumentEnum_ Is Nothing Then

                ' Checking argument count validity
                If pIMCommandArgumentEnum_.ArgsGetNb() > 0 Then

                    ' Checking availability of first argument
                    Dim argument As IMPluginLib.IIMCommandArgument = Nothing
                    pIMCommandArgumentEnum_.ArgGet(0, argument)

                    If Not argument Is Nothing Then

                        ' Checking and retrieving argument value if of valid type
                        If argument.ArgValueGetAsString(1, argValue) = 1 Then

                            success = True

                        End If

                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument)

                    End If

                End If

            End If

            ' Argument value validation
            If success = True Then

                ' TODO: Replace validation code by your own depending on wanted argument value valid range
                If argValue = "Invalid" Then

                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_ARG_INVALID_STRING

                Else

                    ' TODO: Register setting value into the plug-in for future use or execute action using argument value
                    ' TODO: Replace g_stringSettingValue use by your own code
                    Settings.g_stringSettingValue = argValue

                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED

                End If

            End If

            System.Runtime.InteropServices.Marshal.ReleaseComObject(pIMCommandArgumentEnum_)

            Return retCode

        End Function

    End Class

    ' =============================================================================================
    ' TODO: Rename CompanyNameString with your company name and string setting name.
    '
    ' Gets a string value setting from the plug-in
    ' ---------------------------------------------------------------------------------------------
    Public Class CCmdCompanyNameStringGet
        Inherits CComCommand

        ' =========================================================================================
        ' Command constructor. 
        ' This constructor is only called one time when the command is registered to the 
        ' application.
        '
        ' Parameter: info_ : Parameters necessary to build the command.
        ' -----------------------------------------------------------------------------------------
        Public Sub New(ByVal info_ As CComCommandInfo)

            MyBase.New(info_)

        End Sub

        ' =========================================================================================
        ' Gets a string value setting from the plug-in
        '
        ' Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        '
        ' Return: COM error code.
        ' -----------------------------------------------------------------------------------------
        Public Overrides Function Execute2(ByVal pIMCommandArgumentEnum_ As IMPluginLib.IIMCommandArgumentEnum) As IMPluginLib.ECommandReturnCodes

            Dim retCode As IMPluginLib.ECommandReturnCodes = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED

            ' Getting argument value
            If Not pIMCommandArgumentEnum_ Is Nothing Then

                ' Checking argument count validity
                If pIMCommandArgumentEnum_.ArgsGetNb() > 0 Then

                    ' Checking availability of first argument
                    Dim argument As IMPluginLib.IIMCommandArgument = Nothing
                    pIMCommandArgumentEnum_.ArgGet(0, argument)

                    If Not argument Is Nothing Then

                        ' TODO: Get setting value from the plug-in
                        ' TODO: Replace g_stringSettingValue use by your own code
                        If argument.ArgValueSetFromString(1, Settings.g_stringSettingValue) = 1 Then

                            retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED

                        End If

                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument)

                    End If

                End If

            End If

            System.Runtime.InteropServices.Marshal.ReleaseComObject(pIMCommandArgumentEnum_)

            Return retCode

        End Function

    End Class

End Namespace

File: include/PolyWorksSDK/samples/plugins/common/CSharp/CSharpCommands/COM_SDK/CCompanyNameCommands.cs
// =================================================================================================
// Plug-in's commands.
//
// Copyright © InnovMetric Logiciels Inc. 2011 All Rights Reserved
// -------------------------------------------------------------------------------------------------
namespace CSharpCommands
{
    // =============================================================================================
    // Settings
    // =============================================================================================
    public static class Settings
    {
        public static double g_doubleSettingValue  = 110.25;
        public static int    g_integerSettingValue = 237;
        public static string g_stringSettingValue  = "String Value";
    }

    // =============================================================================================
    // TODO: Rename CompanyNameAction with your company name and your action
    //
    // TODO: Add command description here.
    // ---------------------------------------------------------------------------------------------
    public class CCmdCompanyNameAction : IM.CComCommand
    {
        // =========================================================================================
        // Commands constructor. 
        // This constructor is only called one time when the command is registered to the 
        // application.
        //
        // Parameter: info_ : Parameters necessary to build the command.
        // -----------------------------------------------------------------------------------------
        public CCmdCompanyNameAction(IM.CComCommandInfo info_)
            : base(info_)
        {}

        // =========================================================================================
        // Commands action. 
        //
        // TODO: Comment the action here
        //
        // Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        //
        // Return: COM error code.
        // -----------------------------------------------------------------------------------------
        public override IMPluginLib.ECommandReturnCodes Execute2(IMPluginLib.IIMCommandArgumentEnum iIMCommandArgumentEnum_)
        {
            // TODO: Add command execution code here.

            return IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED;
        }
    };

    // =============================================================================================
    // TODO: Rename CompanyNameDouble with your company name and double setting name.
    //
    // Sets a double value setting within the plug-in
    // ---------------------------------------------------------------------------------------------
    public class CCmdCompanyNameDoubleSet : IM.CComCommand
    {
        // =========================================================================================
        // Commands constructor. 
        // This constructor is only called one time when the command is registered to the 
        // application.
        //
        // Parameter: info_ : Parameters necessary to build the command.
        // -----------------------------------------------------------------------------------------
        public CCmdCompanyNameDoubleSet(IM.CComCommandInfo info_)
            : base(info_)
        {}

        // =========================================================================================
        // Sets a double value setting within the plug-in.
        //
        // Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        //
        // Return: COM error code.
        // -----------------------------------------------------------------------------------------
        public override IMPluginLib.ECommandReturnCodes Execute2(IMPluginLib.IIMCommandArgumentEnum iIMCommandArgumentEnum_)
        {
            IMPluginLib.ECommandReturnCodes retCode = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED;

            // Getting argument value
            bool success = false;
            double argValue = 0.0;

            // Setting argument value
            if (iIMCommandArgumentEnum_ != null)
            {
                // Checking argument count validity
                if (iIMCommandArgumentEnum_.ArgsGetNb() > 0)
                {
                    // Checking availability of first argument
                    IMPluginLib.IIMCommandArgument argument = null;
                    iIMCommandArgumentEnum_.ArgGet(0, out argument);
                    if (argument != null)
                    {
                         // Checking and retrieving argument value if of valid type
                        if (argument.ArgValueGetAsDouble(1, out argValue) == 1)
                        {
                            success = true;
                        }
                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument);
                    }
                }
            }

            // Argument value validation
            if (success)
            {
                // TODO: Replace validation code by your own depending on wanted argument value valid range
                if (argValue < 0)
                {
                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_ARG_OUT_OF_RANGE;
                }
                else
                {
                    // TODO: Register setting value into the plug-in for future use or execute action using argument value
                    // TODO: Replace g_doubleSettingValue use by your own code
                    Settings.g_doubleSettingValue = argValue;

                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED;
                }
            }

            System.Runtime.InteropServices.Marshal.ReleaseComObject(iIMCommandArgumentEnum_);

            return retCode;
        }
    };

    // =============================================================================================
    // TODO: Rename CompanyNameDouble with your company name and double setting name.
    //
    // Gets a double value setting from the plug-in
    // ---------------------------------------------------------------------------------------------
    public class CCmdCompanyNameDoubleGet : IM.CComCommand
    {
        // =========================================================================================
        // Commands constructor. 
        // This constructor is only called one time when the command is registered to the 
        // application.
        //
        // Parameter: info_ : Parameters necessary to build the command.
        // -----------------------------------------------------------------------------------------
        public CCmdCompanyNameDoubleGet(IM.CComCommandInfo info_)
            : base(info_)
        {}

        // =========================================================================================
        // Gets a double value setting from the plug-in
        //
        // Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        //
        // Return: COM error code.
        // -----------------------------------------------------------------------------------------
        public override IMPluginLib.ECommandReturnCodes Execute2(IMPluginLib.IIMCommandArgumentEnum iIMCommandArgumentEnum_)
        {
            IMPluginLib.ECommandReturnCodes retCode = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED;

            // Getting argument value
            if (iIMCommandArgumentEnum_ != null)
            {
                // Checking argument count validity
                if (iIMCommandArgumentEnum_.ArgsGetNb() > 0)
                {
                    // Checking availability of first argument
                    IMPluginLib.IIMCommandArgument argument = null;
                    iIMCommandArgumentEnum_.ArgGet(0, out argument);
                    if (argument != null)
                    {
                        // TODO: Get setting value from the plug-in
                        // TODO: Replace g_doubleSettingValue use by your own code
                        if (argument.ArgValueSetFromDouble(1, Settings.g_doubleSettingValue) == 1)
                        {
                            retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED;
                        }
                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument);
                    }
                }
            }

            System.Runtime.InteropServices.Marshal.ReleaseComObject(iIMCommandArgumentEnum_);

            return retCode;
        }
    };

    // =============================================================================================
    // TODO: Rename CompanyNameInteger with your company name and integer setting name.
    //
    // Sets a integer value setting within the plug-in
    // ---------------------------------------------------------------------------------------------
    public class CCmdCompanyNameIntegerSet : IM.CComCommand
    {
        // =========================================================================================
        // Commands constructor. 
        // This constructor is only called one time when the command is registered to the 
        // application.
        //
        // Parameter: info_ : Parameters necessary to build the command.
        // -----------------------------------------------------------------------------------------
        public CCmdCompanyNameIntegerSet(IM.CComCommandInfo info_)
            : base(info_)
        {}

        // =========================================================================================
        // Sets an integer value setting within the plug-in.
        //
        // Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        //
        // Return: COM error code.
        // -----------------------------------------------------------------------------------------
        public override IMPluginLib.ECommandReturnCodes Execute2(IMPluginLib.IIMCommandArgumentEnum iIMCommandArgumentEnum_)
        {
            IMPluginLib.ECommandReturnCodes retCode = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED;

            // Getting argument value
            bool success = false;
            int argValue = 0;

            // Setting argument value
            if (iIMCommandArgumentEnum_ != null)
            {
                // Checking argument count validity
                if (iIMCommandArgumentEnum_.ArgsGetNb() > 0)
                {
                    // Checking availability of first argument
                    IMPluginLib.IIMCommandArgument argument = null;
                    iIMCommandArgumentEnum_.ArgGet(0, out argument);
                    if (argument != null)
                    {
                        // Checking and retrieving argument value if of valid type
                        if (argument.ArgValueGetAsInt(1, out argValue) == 1)
                        {
                            success = true;
                        }
                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument);
                    }
                }
            }

            // Argument value validation
            if (success)
            {
                // TODO: Replace validation code by your own depending on wanted argument value valid range
                if (argValue < 0)
                {
                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_ARG_OUT_OF_RANGE;
                }
                else
                {
                    // TODO: Register setting value into the plug-in for future use or execute action using argument value
                    // TODO: Replace g_integerSettingValue use by your own code
                    Settings.g_integerSettingValue = argValue;

                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED;
                }
            }

            System.Runtime.InteropServices.Marshal.ReleaseComObject(iIMCommandArgumentEnum_);

            return retCode;
        }
    };

    // =============================================================================================
    // TODO: Rename CompanyNameInteger with your company name and integer setting name.
    //
    // Gets a integer value setting from the plug-in
    // ---------------------------------------------------------------------------------------------
    public class CCmdCompanyNameIntegerGet : IM.CComCommand
    {
        // =========================================================================================
        // Commands constructor. 
        // This constructor is only called one time when the command is registered to the 
        // application.
        //
        // Parameter: info_ : Parameters necessary to build the command.
        // -----------------------------------------------------------------------------------------
        public CCmdCompanyNameIntegerGet(IM.CComCommandInfo info_)
            : base(info_)
        {}

        // =========================================================================================
        // Gets a integer value setting within the plug-in.
        //
        // Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        //
        // Return: COM error code.
        // -----------------------------------------------------------------------------------------
        public override IMPluginLib.ECommandReturnCodes Execute2(IMPluginLib.IIMCommandArgumentEnum iIMCommandArgumentEnum_)
        {
            IMPluginLib.ECommandReturnCodes retCode = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED;

            // Setting argument value
            if (iIMCommandArgumentEnum_ != null)
            {
                // Checking argument count validity
                if (iIMCommandArgumentEnum_.ArgsGetNb() > 0)
                {
                    // Checking availability of first argument
                    IMPluginLib.IIMCommandArgument argument = null;
                    iIMCommandArgumentEnum_.ArgGet(0, out argument);
                    if (argument != null)
                    {
                        // TODO: Get setting value from the plug-in
                        // TODO: Replace g_integerSettingValue use by your own code
                        if (argument.ArgValueSetFromInt(1, Settings.g_integerSettingValue) == 1)
                        {
                            retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED;
                        }
                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument);
                    }
                }
            }

            System.Runtime.InteropServices.Marshal.ReleaseComObject(iIMCommandArgumentEnum_);

            return retCode;
        }
    };

    // =============================================================================================
    // TODO: Rename CompanyNameString with your company name and string setting name.
    //
    // Sets a string value setting within the plug-in
    // ---------------------------------------------------------------------------------------------
    public class CCmdCompanyNameStringSet : IM.CComCommand
    {
        // =========================================================================================
        // Commands constructor. 
        // This constructor is only called one time when the command is registered to the 
        // application.
        //
        // Parameter: info_ : Parameters necessary to build the command.
        // -----------------------------------------------------------------------------------------
        public CCmdCompanyNameStringSet(IM.CComCommandInfo info_)
            : base(info_)
        {}

        // =========================================================================================
        // Sets a string value setting within the plug-in.
        //
        // Parameter: iIMCommandArgumentEnum_ : Arguments passed to the command
        //
        // Return: COM error code.
        // -----------------------------------------------------------------------------------------
        public override IMPluginLib.ECommandReturnCodes Execute2(IMPluginLib.IIMCommandArgumentEnum iIMCommandArgumentEnum_)
        {
            IMPluginLib.ECommandReturnCodes retCode = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED;

            // Getting argument value
            bool success = false;
            string argValue = string.Empty;

            // Setting argument value
            if (iIMCommandArgumentEnum_ != null)
            {
                // Checking argument count validity
                if (iIMCommandArgumentEnum_.ArgsGetNb() > 0)
                {
                    // Checking availability of first argument
                    IMPluginLib.IIMCommandArgument argument = null;
                    iIMCommandArgumentEnum_.ArgGet(0, out argument);
                    if (argument != null)
                    {
                        // Checking and retrieving argument value if of valid type
                        if (argument.ArgValueGetAsString(1, out argValue) == 1)
                        {
                            success = true;
                        }
                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument);
                    }
                }
            }

            // Argument value validation
            if (success)
            {
                // TODO: Replace validation code by your own depending on wanted argument value valid range
                if (argValue.CompareTo("Invalid") == 0)
                {
                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_ARG_INVALID_STRING;
                }
                else
                {
                    // TODO: Register setting value into the plug-in for future use or execute action using argument value
                    // TODO: Replace g_stringSettingValue use by your own code
                    Settings.g_stringSettingValue = argValue;

                    retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED;
                }
            }

            System.Runtime.InteropServices.Marshal.ReleaseComObject(iIMCommandArgumentEnum_);

            return retCode;
        }
    };

    // =============================================================================================
    // TODO: Rename CompanyNameString with your company name and string setting name.
    //
    // Gets a string value setting from the plug-in
    // ---------------------------------------------------------------------------------------------
    public class CCmdCompanyNameStringGet : IM.CComCommand
    {
        // =========================================================================================
        // Commands constructor. 
        // This constructor is only called one time when the command is registered to the 
        // application.
        //
        // Parameter: info_ : Parameters necessary to build the command.
        // -----------------------------------------------------------------------------------------
        public CCmdCompanyNameStringGet(IM.CComCommandInfo info_)
            : base(info_)
        {}

        // =========================================================================================
        // Gets a string value setting within the plug-in.
        //
        // Parameter: pIIMCommandArgumentEnum_ : Arguments passed to the command
        //
        // Return: COM error code.
        // -----------------------------------------------------------------------------------------
        public override IMPluginLib.ECommandReturnCodes Execute2(IMPluginLib.IIMCommandArgumentEnum iIMCommandArgumentEnum_)
        {
            IMPluginLib.ECommandReturnCodes retCode = IMPluginLib.ECommandReturnCodes.E_CMD_FAILED;

            // Setting argument value
            if (iIMCommandArgumentEnum_ != null)
            {
                // Checking argument count validity
                if (iIMCommandArgumentEnum_.ArgsGetNb() > 0)
                {
                    // Checking availability of first argument
                    IMPluginLib.IIMCommandArgument argument = null;
                    iIMCommandArgumentEnum_.ArgGet(0, out argument);
                    if (argument != null)
                    {
                        // TODO: Get setting value from the plug-in
                        // TODO: Replace g_stringSettingValue use by your own code
                        if (argument.ArgValueSetFromString(1, Settings.g_stringSettingValue) == 1)
                        {
                            retCode = IMPluginLib.ECommandReturnCodes.E_CMD_SUCCEEDED;
                        }
                        System.Runtime.InteropServices.Marshal.ReleaseComObject(argument);
                    }
                }
            }

            System.Runtime.InteropServices.Marshal.ReleaseComObject(iIMCommandArgumentEnum_);

            return retCode;
        }
    }
}
